### PR TITLE
matrix-synapse: 1.3.1 -> 1.4.0

### PR DESCRIPTION
**Pull request by Vskilet**
###### Motivation for this change
Latest release
https://github.com/matrix-org/synapse/releases/tag/v1.4.0

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Ma27 @pacien

### CODE REVIEW

**Comment by Ma27** [comment]
@GrahamcOfBorg test matrix-synapse

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [inline nixos/modules/services/misc/matrix-synapse.nix]: Please use `mkRemovedOptionModule` here with some instructions what to use instead, otherwise it might be pretty confusing if one option simply went missing :)

Comment (Vskilet) [reply nixos/modules/services/misc/matrix-synapse.nix]: Is that correct ?

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (pacien) [inline nixos/modules/services/misc/matrix-synapse.nix]: What's `#5835`?

Comment (Vskilet) [reply nixos/modules/services/misc/matrix-synapse.nix]: https://github.com/matrix-org/synapse/pull/5875 but I simplify the comment. Do you think it's better ?

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (pacien) [inline nixos/modules/services/misc/matrix-synapse.nix]: Option path element missing, should be:
```suggestion
    (mkRemovedOptionModule [ "services" "matrix-synapse" "trusted_third_party_id_servers" ] ''
```

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (pacien) [reply nixos/modules/services/misc/matrix-synapse.nix]: Just use the release number. It should be enough to suggest the replacement option as in the [changelog](https://github.com/matrix-org/synapse/blob/v1.4.0/CHANGES.md#features).

```suggestion
      Since 1.4.0, "trust_identity_server_for_password_resets" is replaced by "account_threepid_delegates"
```

`account_threepid_delegates` should probably be added to the module.

**Comment by Vskilet** [comment]
Ok pacien I'll create `account_threepid_delegates` option tomorrow

**Comment by risicle** [comment]
Already broken on darwin, no change here unfortunately :(

(macos 10.13)

**Comment by gjabell** [comment]
Any movement on this? Setting up a matrix server soon but I'm waiting for 1.4 to be merged first.

**Comment by Vskilet** [comment]
I'll finished this tonight or tomorrow morning

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Vskilet) [inline nixos/modules/services/misc/matrix-synapse.nix]: Is that an elegant way to this ? @pacien or @Ma27

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Uhm, `trust_identity_server_for_password_resets` != `trusted_third_party_id_servers`. The first one should simply get renamed. Currently working on a fix.

Comment (pacien) [reply nixos/modules/services/misc/matrix-synapse.nix]: Oops, copy/paste error of mine, indeed it's `trusted_third_party_id_servers` instead of `trust_identity_server_for_password_resets`.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (pacien) [reply nixos/modules/services/misc/matrix-synapse.nix]: No, those should be nested using a submodule type: `account_threepid_delegates.email` and `account_threepid_delegates.msisdn`. Take a look at how [`listeners.*` are defined](https://github.com/NixOS/nixpkgs/blob/e2664d0e942236f3244c1e8cefe8c38b8bcb6e29/nixos/modules/services/misc/matrix-synapse.nix#L221).

Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: (currently working on it btw).

**Review by Ma27 — COMMENTED** [review]
So, just did a review and addressed some minor issues. I deployed the current state onto my personal homeserver to check if I encounter any regressions. @pacien @Vskilet please review.

Regarding my changes:

* it seems as `account_threepid_delegates` is simply a block which takes an email and a `msisdn` value (or am I missing something here?)
* Dropped the deprecated `trusted_third_party_id_servers`. People will now get a meaningful error message when using this and upgrading to matrix-synapse 1.4.0.

It would be awesome if the merger of this PR would squash all commits into a single commit with a `Co-Authored-by` statement for @pacien and me.

@NixOS/backports @Ekleog @pacien  I'd propose to backport this even though it contains breaking changes: the old behavior for email activation (which used to rely on `trusted_third_party_id_servers`) will be disabled on hosted identity-servers (matrix.org, vector.im) in December 2019 (https://matrix.org/blog/2019/10/03/synapse-1-4-0-released), so the old behavior can't be supported after that.

Also, people should notice the change as `mkRemovedOptionModule` already raises an evaluation error, right?

**Comment by worldofpeace** [comment]
>  I'd propose to backport this even though it contains breaking changes: the old behavior for email activation (which used to rely on trusted_third_party_id_servers) will be disabled on hosted identity-servers (matrix.org, vector.im) in December 2019 (https://matrix.org/blog/2019/10/03/synapse-1-4-0-released), so the old behavior can't be supported after that.

Correct @Ma27, if they make a decision like this we unfortunately won't have any choice in the matter.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (pacien) [reply nixos/modules/services/misc/matrix-synapse.nix]: It should still be `account_threepid_delegates.email` instead of `account_threepid_delegates_email`. Same for `msisdn`.

**Review by pacien — CHANGES_REQUESTED** [review]

**Comment by Ekleog** [comment]
@Ma27 I didn't have the time to review this yet, so this is neither a r+ or r- and I'll let you handle review, but agree that your arguments appear to me as requiring a backport. Thank you for handling this! :)

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: You're right. Will fix this later today :)

Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Oof, forgot about this, will take care of this tonight :)

**Comment by Ma27** [comment]
@pacien @Ekleog would you mind taking another look? After that I'd merge/backport :)

**Comment by pacien** [comment]
@ofborg test matrix-synapse

**Review by pacien — APPROVED** [review]
Looks good to me.

**Review by Ma27 — APPROVED** [review]
Just squashed the commits into a single one as all changes are actually related (all module changes are needed due to the matrix-synapse bump), so bisecting gets noticeable easier that way. Also added @pacien and myself as co-authors.

I'm running those changes with matrix-synapse 1.4.0 on my personal Matrix setup since Friday without encountering any issues. Also tested/reviewed this with @calbrecht this evening.

I'd wait for ofborg to build and merge/backport (to 19.09) then. IMHO it's fine to not backport to 19.03 as those changes are actually a breaking change and 19.03 support will be dropped before the legacy features are disabled on Matrix.org infrastructure.

@GrahamcOfBorg test matrix-synapse

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Addressed in the latest commit :)

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Fixed in the current changeset :)

**Comment by Ma27** [comment]
Thanks @Vskilet @pacien for the work on this! I just backported this as 1351dde073a5044221fa25ce942dcc7c62269336.

**Pull request merged**
